metricbeat/module/linux/perf/perf.go

```diff
@@ -0,0 +1,277 @@
+// Licensed to Elasticsearch B.V. under one or more contributor
+// license agreements. See the NOTICE file distributed with
+// this work for additional information regarding copyright
+// ownership. Elasticsearch B.V. licenses this file to you under
+// the Apache License, Version 2.0 (the "License"); you may
+// not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+// +build linux
+
+package perf
+
+import (
+	"fmt"
+	"time"
+
+	"github.com/pkg/errors"
+
+	"github.com/elastic/beats/v7/libbeat/common"
+	"github.com/elastic/beats/v7/libbeat/common/cfgwarn"
+	"github.com/elastic/beats/v7/metricbeat/mb"
+	"github.com/elastic/beats/v7/metricbeat/module/linux"
+
+	"github.com/hodgesds/perf-utils"
+)
+
+// init registers the MetricSet with the central registry as soon as the program
+// starts. The New function will be called later to instantiate an instance of
+// the MetricSet for each host defined in the module's configuration. After the
+// MetricSet has been created then Fetch will begin to be called periodically.
+func init() {
+	mb.Registry.MustAddMetricSet("linux", "perf", New)
+}
+
+type eventsConfig struct {
+	HardwareEvents bool `config:"hardware"`
+	SoftwareEvents bool `config:"software"`
+}
+
+type sampleConfig struct {
+	ProcessGlob string       `config:"process"`
+	Events      eventsConfig `config:"events"`
+}
+
+// Config holds the metricset config info for perf
+type Config struct {
+	SamplePeriod time.Duration  `config:"perf.sample_period"`
```

Comment:
You can make sure SamplePeriod config is not zero by adding something like this: SamplePeriod time.Duration  `config:"perf.sample_period" validate:"nonzero,required"`